man/man1/tsf.1
.TH TSF 1NEMO "12 June 2025"

.SH "NAME"
tsf \- type a structured file

.SH "SYNOPSIS"
\fBtsf in=\fPfile [parameter=value] .\|.\|.

.SH "DESCRIPTION"
\fItsf\fP types the contents of a (binary) structured file
(see \fIfilestruct\fP(5NEMO)) in a human readable form
on standard output.
The type, tag, dimensions (if any) and contents of all items are listed.
The contents of \fICharType\fP items are printed directly; 
\fIFloatType\fP and \fIDoubleType\fP
items are printed in floating-point notation; all other types are printed in 
octal for portability reasons (but see \fBoctal=\fP below).
\fIXML\fP format can optionally be selected as well, but this is not been well
exercised. A \fIjson\fP mode has been considered.

.SH "PARAMETERS"
.so man1/parameters
.TP 24
\fBin=\fP\fIfile\fP
Input data is read from \fIfile\fP, which must be a structured binary file.
.br
No default.
.TP
\fBmaxprec=t|f\fP
If true, print float and double item data with maximum precision.
.br
[Default: \fBf\fP].
.TP
\fBmaxline=\fP\fIlines\fP
Limits the number of lines of data typed for each item to \fIlines\fP.
See also \fBallline=\fP below.
.br
[Default: \fB4\fP].
.TP
\fBallline=t|f\fP
Print all lines, it overrides any value of \fBmaxline\fP.
If xml format selected, allline will always be set to t.
.br
[Default: \fBf\fP].
.TP
\fBindent=\fP\fIset-indent\fP
Amount to indent contents of item-sets.
.br
[Default: \fB2\fP].
.TP
\fBmargin=\fP\fIright-margin\fP
Righthand margin for output. Set this to a high value to prevent
possible string problems if file needs to be read by \fIrsf\fP.
.br
[Default: \fB72\fP].
.TP
\fBitem=\fP\fIitem-name\fP
select only this item for output.
.br
Default: all items selected.
.TP
\fBxml=t|f\fP
Output in xml mode? 
If not, old-style human readable tsf mode is selected.
.br
[Default: \fBf\fP].
.TP
\fBoctal=t|f\fP
Output of (short/long) integers in octal? Old versions used to
exchange integer information in octal, we have now switched
by default to decimal (more human readable). By using 
\fBoctal=t\fP you can force this program back in the old
mode. If take output of tsf to be read into rsf, you now
need to force octal=t.
.br
[Default: \fBf\fP].

.SH "DEBUG"
\fBdebug=2\fP reports values of \fIMaxSetLen\fP
as defined in "\fBfilestruct.h\fP".
For \fIMaxTagLen\fP and \fIMaxVecDim\fP see \fIrsf(1NEMO)\fP.

.SH "EXAMPLES"
Here is an N-body snapshot of a Plummer sphere with 10 particles:
.EX

    mkplummer - 10 seed=123 | tsf -

char Headline[28] "init_xrandom: seed used 123"
char History[37] "mkplummer - 10 seed=123 VERSION=3.0c"
set SnapShot
  set Parameters
    int Nobj 10 
    double Time 0.00000 
  tes
  set Particles
    int CoordSystem 66306 
    double Mass[10] 0.100000 0.100000 0.100000 0.100000 0.100000 
      0.100000 0.100000 0.100000 0.100000 0.100000 
    double PhaseSpace[10][2][3] -0.609486 -0.221687 -0.450963 
      -0.0766784 -0.209397 0.396561 4.80925 -0.486182 -1.52575 
      0.418474 0.117766 0.230193 0.545314 0.0795464 0.0632245 
      0.0867019 -0.109128 -0.282036 0.377209 -0.213318 -0.0118330 
      . . .
  tes
tes

.EE

.SH "TLDR"

list brief contents of a binary structured file

% tsf {{path/to/file}}

list more

% tsf {{path/to/file}}

- Show contents of a binary structured file in brief format

tsf {{file}}

- Show contents of a binary structured file, showing all lines, and in full precision

tsf {{file}} True allline=t


.SH "CAVEATS"
The algorithm for implementing \fBmaxline\fP is crude.
.PP
If \fBindent=\fP and \fBmargin=\fP are choosen too large, resp. too
small, they can bump into each other and cause illegable output, which
cannot be read by \fIrsf(1NEMO)\fP either.

.SH "SEE ALSO"
ls(1v), file(1), rsf(1NEMO), csf(1NEMO), qsf(1NEMO), bsf(1NEMO), mkplummer(1NEMO), filestruct(3NEMO), filestruct(5NEMO)

.SH "AUTHOR"
Joshua E. Barnes.

.SH "HISTORY"
.ta +1.25i +4.5i
.nf
xx-xxx-86	V1.0 original version	JEB
xx-apr-88	V2.0 new filestruct package	JEB
9-dec-90	V2.3 helpvec and other minor things	PJT
21-mar-01	V2.7 experimental xml output option	PJT
14-jun-02	V3.0 added octal=, output now decimal	PJT
.fi


src/kernel/io/tsf.c
/*
 * TSF: type the contents of a structured file in human-readable
 *      and possibly abbreviated form.
 * V 1.0: Joshua Barnes 1986	original program development
 *        Peter Teuben 7/87	casting pointers fixed for Ultrix
 * V 1.1  Joshua Barnes 7/87    max-precision option
 * V 1.X  Lyman Hurd 8/87	using deferred input
 * V 2.0  Joshua Barnes 4/88    new filestruct package
 *        Peter Teuben 3/90     made GCC happy by supplying silly return 
 *                                     codes in find_name, find_fmt]
 * V2.1   PJT 12-sep-90		added helpvec to defv
 * V2.2   PJT 5-oct-90		added dprintf() of some system parameters
 * V2.3   PJT 8-dec-90		minor improvements 
 * V2.4   PJT 10-feb-91         added item= keyword
 * V2.5   PJT 12-jun-91		nemo_main, to make it return proper exit()
 * V2.6   pjt 4-mar-94          ansi
 * V2.7   pjt 21-mar-01         XML output option, written in 
 *                              "Coffee Plantation", Tucson, AZ.
 * V3.0   pjt 14-jun-02         integer output not in octal anymore
 * V3.1   pjt 29-aug-02         confirmed bug item=XXX doesn't work anymore
 * V3.1a  pjt 25-nov-03         fix minor (but fatal) xml syntax error
 * V3.1e  pjt 10-aug-09         fix bug handled items > 2GB
 * V3.2   wd  08-sep-11         margin automatic adapts to terminal width
 *                              NOTE:  in a pipe or redirect this is not correct
 * V3.3   pjt 18-sep-2013       fixed the terminal width problem on pipes?
 *            23-oct-2013       another fix, using both stdin and stdout
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <filestruct.h>
#ifdef unix
# include <sys/ioctl.h>
# include <stdio.h>
# include <unistd.h>
#endif

string defv[] = {
    "in=???\n                     input file name ",
    "maxprec=false\n		  print nums with max precision ",
    "maxline=4\n                  max lines per item ",
    "allline=false\n              print all lines (overrides maxline) ",
    "indent=2\n			  indentation of compound items ",
#ifdef unix
    "margin=\n                    righthand margin (default: window width)",
#else
    "margin=72\n		  righthand margin ",
#endif
    "item=\n                      Select specific item",
    "xml=f\n                      output data in XML format? (experimental)",
    "octal=f\n                    Force integer output in octal again?",
    "VERSION=3.3c\n		  3-feb-2014 PJT ",
    NULL,
};

string usage="type contents of a (binary) structured file";

local stream instr;			/* input stream from struct. file   */
local bool maxprec;			/* if true, use max precision forms */
local bool allline;			/* if true, disable maxline limit   */
local int maxline, indent, margin;      /* formating parameters             */
local string testtag;                   /* if non-zero, print only this tag */
local bool xml;                         /* output data in XML format        */
local bool octal;                       /* integer output in octal ?        */

/* local functions */
void   print_item   (string);
void   print_set    (string);
void   print_tes    (string);
void   print_header (string, string, int *);
void   print_data   (string, string, int *);
string find_name    (string);
string find_fmt     (string);
bool   outstr       (string);
void   end_line     (void);
void   change_indent(int);

#define BUFLEN   512


void nemo_main(void)
{
    string *tags;

    dprintf(2,"TSF: MaxSetLen = %d\n",MaxSetLen);   /* DEBUG - PJT */
    instr = stropen(getparam("in"), "r");
    maxprec = getbparam("maxprec");
    maxline = getiparam("maxline");
    indent = getiparam("indent");
#ifdef unix
    if(!hasvalue("margin")) {
      struct winsize w1,w2;
      ioctl(STDIN_FILENO, TIOCGWINSZ, &w1);
      ioctl(STDOUT_FILENO, TIOCGWINSZ, &w2);
      dprintf(1,"w_in=%d w_out=%d\n",w1.ws_col,w2.ws_col);
      if (w1.ws_col==0)
	margin = w2.ws_col;
      else if (w2.ws_col==0)
	margin = w1.ws_col;
      else
	margin = MIN(w1.ws_col,w2.ws_col);
    } else {
      margin = getiparam("margin");
    }
#else
    margin = getiparam("margin");
#endif
    if (margin<10) {
      warning("awfully small margin=%d",margin);
      margin=80;
    }
    if (margin>BUFLEN) {
      warning("BUFLEN %d might be too small (margin=%d)",BUFLEN,margin);
      margin=80;
    }
    dprintf(2,"margin=%d\n",margin);

    allline = getbparam("allline");
    if (hasvalue("item")) warning("item= is broken");
    testtag = getparam("item");
    xml = getbparam("xml");
    if (xml) {     /* if XML output used, it should print all items */
      allline = TRUE;
      /* @todo : should this be added:
         <?xml version='1.0' encoding='UTF-8'?> 
       */
      printf("<nemo>\n");
    }
    octal = getbparam("octal");
    while ((tags = list_tags(instr)) != NULL) {
        print_item(*tags);
	free(*tags);
	free((char *)tags);
    }
    if (xml) printf("</nemo>\n");
}

void print_item(string tag)
{
    string type, *tags, *tp;
    int *dims;

    type = get_type(instr, tag);
    if (streq(type, SetType)) {
	get_set(instr, tag);
	print_set(tag);
	tags = list_tags(instr);
	for (tp = tags; *tp != NULL; tp++)
	    print_item(*tp);
	get_tes(instr, tag);
	print_tes(tag);
	for (tp = tags; *tp != NULL; tp++)
	    free(*tp);
	free((char *)tags);
    } else {
	dims = get_dims(instr, tag);
	print_header(tag, type, dims);
	(void) outstr(" ");
	print_data(tag, type, dims);
	end_line();
	if (dims != NULL)
	    free((char *)dims);
    }
    free(type);
}

void print_set(string tag)
{
    char buf[BUFLEN];  /* danger: buffer overflow */

    if (xml)
      sprintf(buf, "<%s>", tag);
    else
      sprintf(buf, "set %s", tag);
    (void) outstr(buf);
    end_line();
    change_indent(indent);
}
    
void print_tes(string tag)
{
    char buf[BUFLEN];  /* danger: buffer overflow */

    change_indent(- indent);
    if (xml)
      sprintf(buf, "</%s>",tag);
    else
      sprintf(buf, "tes");
    (void) outstr(buf);
    end_line();
}

void print_header(string tag, string type, int *dims)
{
    char buf[BUFLEN];  /* danger: buffer overflow */
    int *dp;
    
    if (xml)
      sprintf(buf,"<%s type=\"%s\" ",tag,find_name(type));
    else
      sprintf(buf, "%s %s", find_name(type), tag);
    (void) outstr(buf);
    if (dims != NULL) {				/* is this a plural item?   */
      if (xml) {
	sprintf(buf,"dim=\"");
        outstr(buf);
      }
      for (dp = dims; *dp != 0; dp++) {	/*   loop over dimensions   */
	sprintf(buf, "[%d]", *dp);		/*     format a dimension   */
	(void) outstr(buf);                 /*     and print it out     */
      }
      if (xml) {
	sprintf(buf,"\"");
        outstr(buf);
      }
    }
    if (xml) {
      sprintf(buf,">");
      outstr(buf);
    }
    
}

void print_data(string tag, string type, int *dims)
{
    string fmt;
    size_t dlen;
    char buf[BUFLEN];  /* danger: buffer overflow */
    byte *dat, *dp;
    bool Qprint;

    Qprint = !(*testtag && streq(tag,testtag)==0);  /* if only print testtag... */

    fmt = find_fmt(type);
    dlen = get_dlen(instr, tag);
    dat = (byte*) allocate(dlen+1);
    get_data_sub(instr, tag, type, dat, dims, FALSE);
    if (streq(type, CharType) && Qprint)
	(void) outstr(dims != NULL ? "\"" : "\'");
    for (dp = dat; dp < dat + dlen; ) {		/* loop over data array     */
	if (streq(type, AnyType)) {		/*   output generic data?   */
	    sprintf(buf, fmt, *((byte *) dp));
	    dp += sizeof(byte);
	} else if (streq(type, CharType)) {	/*   output readable chars? */
	    sprintf(buf, fmt, *((char *) dp));
	    dp += sizeof(char);
	} else if (streq(type, ByteType)) {	/*   output bytes of data?  */
	    sprintf(buf, fmt, *((byte *) dp));
	    dp += sizeof(byte);
	} else if (streq(type, ShortType)) {	/*   output short integers? */
	    sprintf(buf, fmt, *((short *) dp));
	    dp += sizeof(short);
	} else if (streq(type, IntType)) {	/*   output standard ints?  */
	    sprintf(buf, fmt, *((int *) dp));
	    dp += sizeof(int);
	} else if (streq(type, LongType)) {	/*   output long integers?  */
	    sprintf(buf, fmt, *((long *) dp));
	    dp += sizeof(long);
	} else if (streq(type, HalfpType)) {	/*   output short int? */
	    sprintf(buf, fmt, *((short *) dp));
	    dp += sizeof(short);
	} else if (streq(type, FloatType)) {	/*   output floating point? */
	    sprintf(buf, fmt, *((float *) dp));
	    dp += sizeof(float);
	} else if (streq(type, DoubleType)) {	/*   output double numbers? */
	    sprintf(buf, fmt, *((double *) dp));
	    dp += sizeof(double);
	} else
	    error("print_data: type %s unknown\n", type);
	if (Qprint)
	  if (! outstr(buf))
	    break;
    }
    if (streq(type, CharType) && Qprint)
	(void) outstr(dims != NULL ? "\"" : "\'");
    if (xml && Qprint) {
      sprintf(buf," </%s>",tag);
      outstr(buf);
    }
    free((char *)dat);
}

/*
 * TYPEINFO: struct for info about data types.
 */

struct typeinfo {
    string titype;		/* type code used by filestruct package     */
    string tiname;		/* C-style declarator for this type         */
    string tilpfmt;		/* printf format for low-precision output   */
    string tihpfmt;		/* printf format for max-precision output   */
    string tixpfmt;             /* printf format for XML output */
};

/*
 * TYPETAB: table of info about data types.
 */

struct typeinfo typetab_o[] = {
    { AnyType,    "any",      "%#o ",     "%#o ",     "%#o", },
    { CharType,   "char",     "%c",       "%c",       "%c",  },
    { ByteType,   "byte",     "%#o ",     "%#o ",     "%#d ", },
    { ShortType,  "short",    "%#o ",     "%#o ",     "%#d ", },
    { IntType,    "int",      "%#o ",     "%#o ",     "%#d ", },
    { LongType,   "long",     "%#lo ",    "%#lo ",    "%#d ", },
    { HalfpType,  "halfp",    "%#o ",     "%#o ",     "%#d ", },
    { FloatType,  "float",    "%#g ",     "%15.8e ",  "%#g ", },
    { DoubleType, "double",   "%#g ",     "%23.16e ", "%#g ", },
    { SetType,    "set",      NULL,       NULL,        NULL, },
    { TesType,    "tes",      NULL,       NULL,        NULL, },
    { NULL, },
};


struct typeinfo typetab_d[] = {
    { AnyType,    "any",      "%#o ",     "%#o ",     "%#o", },
    { CharType,   "char",     "%c",       "%c",       "%c",  },
    { ByteType,   "byte",     "%#d ",     "%#d ",     "%#d ", },
    { ShortType,  "short",    "%#d ",     "%#d ",     "%#d ", },
    { IntType,    "int",      "%#d ",     "%#d ",     "%#d ", },
    { LongType,   "long",     "%#ld ",    "%#ld ",    "%#d ", },
    { HalfpType,  "halfp",    "%#d ",     "%#d ",     "%#d ", },
    { FloatType,  "float",    "%#g ",     "%15.8e ",  "%#g ", },
    { DoubleType, "double",   "%#g ",     "%23.16e ", "%#g ", },
    { SetType,    "set",      NULL,       NULL,        NULL, },
    { TesType,    "tes",      NULL,       NULL,        NULL, },
    { NULL, },
};

string find_name(string type)
{
    struct typeinfo *tip;

    for (tip = typetab_o; tip->titype != NULL; tip++)
	if (streq(tip->titype, type))
	    return (tip->tiname);
    error("find_name: type %s unknown\n", type);
    return NULL;   /* make stringent compilers happy */
}

string find_fmt(string type)
{
    struct typeinfo *tip;

    for (tip = octal ? typetab_o : typetab_d; tip->titype != NULL; tip++)
	if (streq(tip->titype, type))
	  return (xml ? tip->tixpfmt :
		  maxprec ? tip->tihpfmt : tip->tilpfmt);
    error("find_fmt: type %s unknown\n", type);
    return NULL;   /* make stringent compilers happy */
}

/*
 * State variables for formatting functions.
 */

int curcoll = 0;	/* current cursor column, numberd from 0 */
int curline = 0;	/* current line of item, numbered from 0 */

int baseindent = 0;	/* indentation of item */
int actindent;		/* actual indentation of line */

bool outstr(string str)
{
    if (curcoll + (int)strlen(str) >= margin) { /* string too long? */
	printf("\n");				/*   begin next line */
	curcoll = 0;
	curline++;
	if ((!allline) && (curline == maxline))	/*   too much output? */
	    str = ". . .";
	actindent = baseindent + indent;	/*   indent text of item */
    }
    while (curcoll < actindent) {		/* space over to start */
	printf(" ");
	curcoll++;
    }
    printf("%s", str);                          /* output string */
    curcoll = curcoll + strlen(str);            /* and count length */
    return (allline || (curline < maxline));
}

void end_line(void)
{
    printf("\n");
    curcoll = curline = 0;
    actindent = baseindent;
}

void change_indent(int delta)
{
    baseindent = baseindent + delta;
    actindent = baseindent;
}
